docs/adrs/002-Retry_endpoint.md

```diff
@@ -0,0 +1,87 @@
+* Link to Jira Ticket with business logic documentation for success and cancel cases
+
+### Possible additional things to think about
+* Creating sets of objects and what to do at the end of a batch run
+* What to do about a set if a child task fails during processing 
+* Convenience endpoints that operate on the parent_job_id or list of child job ids may be out of scope (e.g. cancel all jobs with a certain status)
+
+## Consequences
+
+Pending decision
+
+## Pros and Cons of the Alternatives
+
+### Not book-keeping, or doing minimal book-keeping and calling `run_job` multiple times
+* `+` Can re-use existing endpoints without any additional work
+* `+` Less api endpoints to manage
+* `-` Issues on re-rendering/regenerating a cell based on just the job record
+* `-` Loss of information about job runs, and ability to infer relationships between parents and child jobs. 
+* `-` Loss of control of jobs, such as the ability to restrict a job's running based on number of retries/failures.
+* `-` Wrong version of app will run if the app was updated after job completion, and a version tag rather than a git commit was provided
+* `-` Increase complexity of `run_job*` methods
+* `-` The client will have to keep track of the child_job relationship, so that info is lost once the client is terminated
+ 
+### Re-writing `run_job` to address the aforementioned deficiencies without refactoring
+* `+` Solves most requirements, but 
+* `-` Adds more complexity to `run_job` methods
+* `-` Increase difficulty in maintaining and testing `run_job` method
+* `-` Wrong version of app will run if the app was updated after job completion, and a version tag rather than a git commit was provided
+* `-` Inefficient job submission
+* `-` Possibly Insufficient error handling 
+
+### Re-writing `run_job/run_job_batch` to address the aforementioned deficiencies with some refactoring
+* `+` Same as above, but if you are refactoring, you might as well have a retry endpoint, and clean out/decouple `run_job` endpoint from having so many features and branching logic
+
+### Creating `retry` endpoint to address the aforementioned deficiencies with some refactoring
+* `+` Decrease coupling between `run_job` and retry functionality, possibly making testing and development easier
+* `+` Faster development than a full refactor
+* `-` Faster development than a full refactor, but creates technical debt, might have to update both `run_job` and `retry` each time a change is made
+* `-` Extra endpoint to manage
+
+
+### Creating `retry` endpoint to address the aforementioned deficiencies with full refactoring where run_job functions are split out into their own functions
+* `+` Decrease coupling between `run_job` and retry functionality, possibly making testing and development easier
+* `+` Increase DRYNESS of the code
+* `+` Allows retry to benefit from changes to `run_job` 
+* `-` Slower development for a full refactor, but decreases technical debt
```

Comment:
Before we make a decision on implementation details like refactoring vs no refactoring, I feel like we should figure out the behavior we want a job retry to have. AJ brought up some good points here: https://github.com/kbase/execution_engine2/pull/367/files#r621719800

Behavior questions, incorporating AJ's notes and my own (AJ let me know if I missed any):

* Should we track a retry count?
* Should users see this retry count?
* Are retried jobs saved in some sort of data structure linking them, possibly indirectly, to the parent job or are they orphaned?
* If the former, is the retry relationship linear or a tree? E.g. what happens if there are two simultaneous calls to retry a job?
* Should it be at least theoretically possible to see the list of retried jobs in order?
* Should there be a maximum retry count? Or a warning that more retries are not likely to help?
* Can a job in states other than failed or canceled be retried? Or should the user be required to cancel a job before it can be retried?

Reply:
Once we make those decisions, to me the next step would be figuring out the data model in the EE2 DB we'd want to use to represent the state of retried jobs and how to make changes to that model atomically, or if that's not possible, to make the model resilient to unexpected server / network / db outages. Next the operations on that model that occur during a retry. That'll tell us how much would have to change in the EE2RunJob class, assuming that handles retries as well.

Reply:
Does this make sense to people or am I off base?

Reply:
I think these questions are better suited on the retry ticket. From my understanding, the point of this ADR was to justify a retry endpoint, and some of the reasons it better to have that then to use the existing run_job endpoint as is.

Reply:
My understanding was that we had decided we *did* want a retry endpoint, and the purpose of the ADR was to figure out how it should work - that's what the notes from the engineering checkin say anyway.

Reply:
I am with Gavin on this, we are agreed on the desirability of the retry endpoint, but want to document the decision and discuss the details of what the retry endpoint will do, and discuss the effort involved for different options.

As I mentioned, as the product owner, I find our ability to deliver functionality to be pretty awful. We have invested so much effort in refactoring that its killed our timeline - we started in late July, and it is now almost May with no functioning bulk uploader, which was just the first deliverable.

If we are going to refactor, we need to be able to do it in a timely fashion, and have it not kill the schedule any more than it has. I want to see the estimate for a quick and dirty solution that implements a proposed retry endpoint, that can be deployed ASAP, and then once the API contract has been established, and the functional MVP is done, we begin the cleanup of the backend code.

Note that this is *NOT* business as usual, the usual way we do this is the nasty MVP gets deployed and then we don't go back until much later. Here, we get the API working so that it doesn't block dependencies, and we immediately start the refactoring. The refactor needs to be broken down into smallish chunks of ~3 days estimated work, and each merge should maintain functionality and incrementally improve the codebase. Tasks that take more than a couple of days are more likely to be far off in their estimate and this is how we mitigate the risk of poor estimation.

Reply:
> Should it be at least theoretically possible to see the list of retried jobs in order?

If this is not done directly, it could be done by ordering the jobs by creation time

Reply:
> Should there be a maximum retry count? Or a warning that more retries are not likely to help?

This is something that could be enforced by the client if it isn't implemented by ee2.

Reply:
> Are retried jobs saved in some sort of data structure linking them, possibly indirectly, to the parent job or are they orphaned?
> If the former, is the retry relationship linear or a tree? E.g. what happens if there are two simultaneous calls to retry a job?

If the retried jobs are not linked to the parent job, we lose a lot of the power of having a retry endpoint. To simplify things, I think it would be best for only one execution of a given set of parameters to be permitted at a time. This implies that ee2 is able to distinguish unique jobs in a batch; the simplest way to do this would be to generate a hash from each set of params [or, if there is a unique param, such as output object name, use that], and reject any duplicated param sets.

Another possible alternative would be tracking the position of the job in the initial input set and having a field that links the rerun to the original job.